rust_programs/ttf_renderer/src/glyphs.rs

```diff
@@ -562,3 +573,107 @@ fn interpret_values_via_flags(
     }
     values
 }
+
+fn bezier_2d(start: isize, ctrl: isize, end: isize, t: f64) -> isize {
+    return ((start as f64) + (2 as f64)*t*(ctrl as f64) + t*t*((end - ctrl) as f64)).round() as isize;
+}
+
+fn points_for_polygon(
+    start_index: usize,
+    end_index: usize,
+    points: &Vec<isize>,
+    flags: &Vec<Vec<GlyphOutlineFlag>>,
+    interpolate: bool
+) -> Vec<isize> {
+    
+    let mut values = vec![];
+
+    if !interpolate {
+        return points[start_index..=end_index].to_vec().clone();
+    }
+    
+    /*
+    for index in start_index..=end_index {
+        values.push(points[index]);
+    }
+    return values;
+    */
+
+    if start_index == end_index {
+        return values;
+    }
+
+    let mut last_value = points[start_index];
+    values.push(last_value);
+    
+    let mut last_on_curve_value = last_value;
+    let mut last_off_curve_value = 0;
+    let mut last_value_was_on_curve = true;
+
+    
+    for index in start_index+1..=end_index {
+        let next_value = points[index];
+        let relative_value = next_value - last_value;
+        last_value = next_value;
+        let flag_set = &flags[index];
+            
+        let next_pt_is_on_curve = flag_set.contains(&GlyphOutlineFlag::OnCurve);
+
+        if last_value_was_on_curve {
+            if next_pt_is_on_curve {
+                // just add new point to the list of points (i.e. draw a
+                // straight line to the new point)
+                last_on_curve_value = next_value;
+                values.push(last_on_curve_value);
+            }
+            else {
+                // the next point is a control point, so we have a 2nd degree
+                // bezier curve, but we don't know the end point yet. Store
+                // the relative value for next round
+                last_off_curve_value = relative_value;
+                last_value_was_on_curve = false;
```

Comment:
No need to change this, just a Rust FYI: Rather than managing multiple state variables to point to the same 'conceptual' data, you can use a sum type to track both the optionality and data at the same time:

```
// Before
let mut is_foo_valid = false;
let mut foo = 0;
...
foo = 123;
is_foo_valid = true;
...
if is_foo_valid {
    do_something_with_foo(foo);
}

// After
let mut maybe_foo = None;
...
maybe_foo = Some(123);
...
if let Some(foo) = maybe_foo {
    do_something_with_foo(foo)
}
```

Reply:
Ended up implementingt this as well, since it was related to the next comment